cleverhans/attacks_tf.py

```diff
@@ -257,151 +184,138 @@ def jacobian_graph(predictions, x, nb_classes):
+            list_derivatives.append(derivatives[0])
+        grads = tf.reshape(tf.stack(list_derivatives),
+                           shape=[nb_classes, -1, nb_features])
 
         # Compute the Jacobian components
-        grads_target, grads_others = jacobian(sess, x, grads, target,
-                                              adv_x_original_shape,
-                                              nb_features, nb_classes,
-                                              feed=feed)
-
-        if iteration % ((max_iters + 1) // 5) == 0 and iteration > 0:
-            _logger.debug("Iteration {} of {}".format(iteration,
-                                                      int(max_iters)))
-        # Compute the saliency map for each of our target classes
-        # and return the two best candidate features for perturbation
-        i, j, search_domain = saliency_map(
-            grads_target, grads_others, search_domain, increase)
-
-        # Apply the perturbation to the two input features selected previously
-        adv_x = apply_perturbations(
-            i, j, adv_x, increase, theta, clip_min, clip_max)
-
-        # Update our current prediction by querying the model
-        current = utils_tf.model_argmax(sess, x, predictions,
-                                        adv_x_original_shape, feed=feed)
+        target_class = tf.reshape(tf.transpose(y_in, perm=[1, 0]),
+                                  shape=[nb_classes, -1, 1])
+        other_classes = tf.cast(tf.not_equal(target_class, 1), tf.float32)
+
+        grads_target = tf.reduce_sum(grads * target_class, axis=0)
+        grads_other = tf.reduce_sum(grads * other_classes, axis=0)
+
+        # Remove the already-used input features from the search space
+        # Subtract 2 times the maximum value from those value so that
+        # they won't be picked later
+        increase_coef = (4 * int(increase) - 2) \
+            * tf.cast(tf.equal(domain_in, 0), tf.float32)
+
+        target_tmp = grads_target
+        target_tmp -= increase_coef \
+            * tf.reduce_max(tf.abs(grads_target), axis=1, keep_dims=True)
+        target_sum = tf.reshape(target_tmp, shape=[-1, nb_features, 1]) \
+            + tf.reshape(target_tmp, shape=[-1, 1, nb_features])
+
+        other_tmp = grads_other
+        other_tmp += increase_coef \
+            * tf.reduce_max(tf.abs(grads_other), axis=1, keep_dims=True)
+        other_sum = tf.reshape(other_tmp, shape=[-1, nb_features, 1]) \
+            + tf.reshape(other_tmp, shape=[-1, 1, nb_features])
+
+        # Create a mask to only keep features that match conditions
+        if increase:
+            scores_mask = ((target_sum > 0) & (other_sum < 0))
+        else:
+            scores_mask = ((target_sum < 0) & (other_sum > 0))
 
-        # Update loop variables
-        iteration = iteration + 1
+        # Create a 2D numpy array of scores for each pair of candidate features
```

Comment:
Quick question: suppose we have an N by N image and we need to find the best pair to perturb, according to this code, we need a matrix of N^2 by N^2, am I understanding it right? In this case, suppose we want to perturb an image of size100x100, then we need approx. 2G memory.

Reply:
Yes, you are right. And it will take more memory if the image has multiple channels.